### PR TITLE
`organizeDeclarations` custom naming & ordering

### PR DESCRIPTION
**Pull request by oiuhr**
This PR adds an ability to parametrize order inside `organizeDeclarations` categories & their respective marks.

I'm not sure about the design so it's up to discussion, but long story short:

We have 6 visibility categories (plus 2 for airbnb conventions, but i guess this number can increase without any major changes) and ~15 type groups. By providing parameters `--visibilityorder` we sort visibility categories (the same goes for the `typeorder`), and by `--mode` parameter we tell which category will be respected as primary sorting group. Each group either take a default `mark comment` value or custom parametrized with parameters `--visibilitymark` & `--typemark`, if provided.

### CODE REVIEW

**Comment by oiuhr** [comment]
Current questions:

- Is it okay if we cant filter out custom marks? For example: we were using default mark for instance lifecycle group, which was given an `// MARK: Lifecycle ` mark. Then we change up to custom mark, smth like `// MARK: Init` and `// MARK: Lifecycle ` is replaced. But if we jump from custom to default formatter simply have nothing to compare and just place `// MARK: Init` under previously used custom mark. Should we add something like `--potentialCategoryMarks` to filter them out if needed?
- I couldn't find the way to somehow validate provided parameters. For example `--visibilityorder` clearly should have 6+ entries in it, all of which should be unique and valid 💀

File: Sources/Examples.swift [hunk 1]
```diff
@@ -1288,6 +1288,42 @@ private struct Examples {
...
+        [
+            beforeMarks,
+            instanceLifecycle,
+            open,
+            public,
+            package,
+            internal,
+            fileprivate,
+            private
+        ]
```
Comment (calda) [inline Sources/Examples.swift:1302]: Since I assume the argument itself doesn't allow the `[` and `]` characters, maybe we should omit them from the default example? 

```suggestion
        beforeMarks,
        instanceLifecycle,
        open,
        public,
        package,
        internal,
        fileprivate,
        private
```

File: Sources/FormattingHelpers.swift [hunk 2]
```diff
@@ -1905,19 +1927,21 @@ extension Formatter {
             }
         }
 
-        func shouldBeMarked(in declarationTypes: [DeclarationType]) -> Bool {
+        func shouldBeMarked(in _: [DeclarationType]) -> Bool {
```
Comment (calda) [inline Sources/FormattingHelpers.swift:1930]: If this argument is unused let's clean it up

```suggestion
        func shouldBeMarked() -> Bool {
```

File: Sources/Examples.swift [hunk 3]
```diff
@@ -1288,6 +1288,42 @@ private struct Examples {
...
+    ```
+
+    default value for `--typeorder`:
```
Comment (calda) [inline Sources/Examples.swift:1305]: We should probably add a comment to `enum DeclarationType` along the lines of:

```
/// Note: When adding a new declaration type, remember to also update the list in `Examples.swift`
```

File: Sources/FormattingHelpers.swift [hunk 4]
```diff
@@ -1832,16 +1833,35 @@ extension Formatter {
...
         case visibility(Visibility)
         case explicit(DeclarationType)
 
+        init?(rawValue: String) {
+            if let visibility = Visibility(rawValue: rawValue) {
+                self = .visibility(visibility)
+            } else if let type = DeclarationType(rawValue: rawValue) {
+                self = .explicit(type)
+            } else {
+                return nil
```
Comment (calda) [inline Sources/FormattingHelpers.swift:1846]: > I couldn't find the way to somehow validate provided parameters. For example --visibilityorder clearly should have 6+ entries in it, all of which should be unique and valid 💀

Perhaps we could do a bit of validation here, and `assertionFailure` if the value can't be parsed.

File: Sources/OptionDescriptor.swift [hunk 5]
```diff
@@ -869,6 +869,30 @@ struct _Descriptors {
...
+    let typeOrder = OptionDescriptor(
+        argumentName: "typeorder",
+        displayName: "Organization Order For Declaration Types",
+        help: "Order for declaration type groups inside declaration",
+        keyPath: \.typeOrder
+    )
+    let customVisibilityMarks = OptionDescriptor(
+        argumentName: "visibilitymarks",
+        displayName: "Custom Visibility Marks",
+        help: "Mark for concrete visibility group (public_Public Fields)",
```
Comment (calda) [inline Sources/OptionDescriptor.swift:887]: I've used the `:` syntax as a separator in this sort of mapping argument before. I think that would be a bit better than an underscore. Also, including `,...` could be helpful here to indicate that it supports multiple values.

```suggestion
        help: "Mark for concrete visibility group (public:Public Fields,...)",
```

File: Sources/OptionDescriptor.swift [hunk 6]
```diff
@@ -869,6 +869,30 @@ struct _Descriptors {
...
+        keyPath: \.organizationMode
+    )
+    let customTypeMarks = OptionDescriptor(
+        argumentName: "typemarks",
+        displayName: "Custom Type Marks",
+        help: "Mark for concrete declaration type group (classMethod_Baaz)",
```
Comment (calda) [inline Sources/OptionDescriptor.swift:893]: Same here

```suggestion
        help: "Mark for concrete declaration type group (classMethod:Baaz,...)",
```

File: Sources/Options.swift [hunk 7]
```diff
@@ -772,6 +776,36 @@ public struct FormatOptions: CustomStringConvertible {
                 organizeEnumThreshold: Int = 0,
                 organizeExtensionThreshold: Int = 0,
                 organizationMode: DeclarationOrganizationMode = .visibility,
+                visibilityOrder: [String] = [
+                    "beforeMarks",
```
Comment (calda) [inline Sources/Options.swift:780]: Could we define this default value as a static property on the `Visibility` type over in the main file? That would help us avoid forgetting to update this

File: Sources/Options.swift [hunk 8]
```diff
@@ -772,6 +776,36 @@ public struct FormatOptions: CustomStringConvertible {
...
+                    "instanceLifecycle",
+                    "open",
+                    "public",
+                    "package",
+                    "internal",
+                    "fileprivate",
+                    "private",
+                ],
+                typeOrder: [String] = [
```
Comment (calda) [inline Sources/Options.swift:789]: Same here, could we define this default ordering as a static property on the `DeclarationType` type in the main file? Keeping things centrally located could help us avoid forgetting to update it when adding a new declaration type.

File: Tests/RulesTests+Organization.swift [hunk 9]
```diff
@@ -384,7 +384,15 @@ class OrganizationTests: RulesTests {
...
-            // MARK: Content
+            // MARK: Content Properties
+
+            @ViewBuilder
+            private var toggle: some View {
+                Toggle(label, isOn: $isOn)
+                    .fixedSize()
+            }
+
+            // MARK: Content Methods
```
Comment (calda) [inline Tests/RulesTests+Organization.swift:395]: Is this change necessary? I liked the previous design with a single `// MARK: Content` more.

**Comment by calda** [comment]
> Is it okay if we cant filter out custom marks? For example: we were using default mark for instance lifecycle group, which was given an // MARK: Lifecycle  mark. Then we change up to custom mark, smth like // MARK: Init and // MARK: Lifecycle  is replaced. But if we jump from custom to default formatter simply have nothing to compare and just place // MARK: Lifecycle under previously used custom mark. Should we add something like --potentialCategoryMarks to filter them out if needed?

To me it seems fine to leave this as-is.

**Review by calda — APPROVED** [review]
Really cool, nice work! The implementation is really simple. Great to see how this builds on top of #1678.

I left some minor comments but otherwise this looks great to me!

File: Sources/FormattingHelpers.swift [hunk 10]
```diff
@@ -1933,21 +1957,90 @@ extension Formatter {
...
+
+    typealias ParsedOrder = [Category]
+    func categoryOrder(for mode: DeclarationOrganizationMode) -> ParsedOrder {
+        typealias ParsedVisibilityMarks = [VisibilityType: String]
+        typealias ParsedTypeMarks = [DeclarationType: String]
+
+        let visibilityTypes = options.visibilityOrder
+            .map { VisibilityType(rawValue: $0) }
+            .compactMap { $0 }
+        let declarationTypes = options.typeOrder
```
Comment (calda) [inline Sources/FormattingHelpers.swift:1971]: To properly support this, I think we'll also need to tweak the implementation of `type(of declaration:for:) -> DeclarationType`.

Some issues that come to mind:
 - If a type of declaration is not included in the user-specified declaration types, we probably need to not create instances of it in `type(of:declaration:for:)`. Otherwise, declarations of that type will still be sorted based on that type, even if is meant to be disabled.
 - Declaration types like `swiftUIProperty` and `swiftUIMethod` are currently disabled when using the `visibility` declaration mode. This makes sense for the Airbnb-style default behavior, but would seem broken for users trying to include it in their custom declaration ordering.

So, I think we need a more flexible implementation of `type(of:declaration:for:)` that, for each declaration type, checks to see if it is enabled for the given mode before generating instances of it.

File: Sources/Options.swift (hunk 7 above)
Comment (calda) [reply Sources/Options.swift:780]: Or even better, the default could just be `nil ` (`visibilityOrder: [String]? = nil`) and we could generate the default ordering using `CaseIterable` like we were doing before.

That sounds even better, because that avoids having multiple sources of truth that have to be kept in sync.

File: Sources/Options.swift (hunk 8 above)
Comment (calda) [reply Sources/Options.swift:789]: Or this other approach would be even better: https://github.com/nicklockwood/SwiftFormat/pull/1736/files#r1644509140

File: Sources/FormattingHelpers.swift (hunk 10 above)
Comment (calda) [reply Sources/FormattingHelpers.swift:1971]: It would be good to see some test cases covering this as well.

File: Tests/RulesTests+Organization.swift [hunk 11]
```diff
@@ -410,6 +412,274 @@ class OrganizationTests: RulesTests {
...
+        }
+        """
+
+        testFormatting(
+            for: input, output,
+            rule: FormatRules.organizeDeclarations,
+            options: FormatOptions(
+                organizationMode: .type,
+                visibilityOrder: ["private", "internal", "public"],
+                typeOrder: ["instanceLifecycle", "instanceMethod", "instanceProperty", "overriddenMethod"]
```
Comment (calda) [inline Tests/RulesTests+Organization.swift:595]: Same for this part -- what happens if the code includes declarations with types that are not included in the `typeOrder`? Would like to see tests covering that.

File: Tests/RulesTests+Organization.swift [hunk 12]
```diff
@@ -410,6 +412,274 @@ class OrganizationTests: RulesTests {
...
+            override public func baar()
+        }
+        """
+
+        testFormatting(
+            for: input, output,
+            rule: FormatRules.organizeDeclarations,
+            options: FormatOptions(
+                organizationMode: .type,
+                visibilityOrder: ["private", "internal", "public"],
```
[the remark this answers is not in the document]
Comment (oiuhr) [reply Tests/RulesTests+Organization.swift:594]: rn it will just throw at runtime due to `!` optional unwrap at line 2043 of `FormattingHelpers` — that's the main reason why i'm asking about parameter validation. same goes for unhandled `typeorder` parameters — if we cant find it in user defined `order`, than we simply can't define where to put it

File: Sources/FormattingHelpers.swift (hunk 10 above)
Comment (oiuhr) [reply Sources/FormattingHelpers.swift:1971]: > If a type of declaration is not included in the user-specified declaration types, we probably need to not create instances of it in type(of:for:)

can you provide some examples for reference, please? as right now, I see your take as `if user did not specify any 'typerorder' — were not sorting` 💥

Comment (calda) [reply Sources/FormattingHelpers.swift:1971]: Here are two scenarios that I think it's important for us to handle:

 1. Imagine in the future we add a new `DeclarationType` (an an example, `propertyWithPropertyWrapper`). For any existing consumers of the `typeOrder`, SwiftFormat will now crash when formatting their code: their configuration doesn't specify `propertyWithPropertyWrapper`, because it didn't previously exist, but we're now generating instances of it. I think we definitely need to handle this more gracefully.

 2. Imagine the user doesn't want to overridden methods differently from other methods. If they omit the `overriddenMethod` declaration type, we should avoid generating instances of it and instead generate something like `instanceMethod`. The same applies to `swiftUIMethod` -- if that type isn't enabled, we should fall back to generating `instanceMethod`.

I understand the current implementation requires generating a `DeclarationType` for every declaration. Maybe we add an `unclassified` case as a fallback for cases where the user-specified order is missing something important (e.g. a static method, but the user's `typeorder` configuration doesn't include `staticMethod`).

Alternatively, we could do something like:
 1. Generate the `DeclarationType` using the user-specified `typeorder` list, making sure we don't generate any `DeclarationType` that isn't in the list. This handles the case where a user may want to disable the `overriddenMethod` or `swiftUIMethod` cases.
 2. If that returns `nil`, because the user's `typeorder` isn't exhaustive, we could re-generate the `DeclarationType` using the full set of declaration types. This is guaranteed to return a valid `DeclarationType`, and may be a better user experience than the other options.

Comment (oiuhr) [reply Sources/FormattingHelpers.swift:1971]: great design, thank you!

i guess we have only a few `required` types (such as `lifecycle (init's)`, 'method', 'property' and 'beforeMarks') which `always` should be defined (and i suppose this could be validated) and everything else is just a co-variant to types above. I'll try to come to something like:

- always specify ALL of visibility types (6 + smth like airbnb conventions)
- always specify 4 required types (`lifecycle`, 'method', 'property' and 'beforeMarks') + some unfolding to them such as `class method` and `static property`

this will assure (to some point) un-need for something like `unclassified` and to some degree constraint rule and its parameters. seems ok?

File: Sources/Examples.swift (hunk 1 above)
Comment (oiuhr) [reply Sources/Examples.swift]: ✅

File: Sources/Examples.swift (hunk 3 above)
Comment (oiuhr) [reply Sources/Examples.swift]: thanks for an example!

File: Sources/FormattingHelpers.swift (hunk 4 above)
Comment (oiuhr) [reply Sources/FormattingHelpers.swift:1878]: removed `assertionFailure` from `init` and added full validation to `OptionDescriptor`

File: Tests/RulesTests+Organization.swift (hunk 9 above)
Comment (oiuhr) [reply Tests/RulesTests+Organization.swift:399]: yeah, they should be separated imo. by the way you can achieve previous results by just removing `contentProperties` from `typeOrder` and rename `contentMethods` category mark to `Content` 🫡

File: Tests/RulesTests+Organization.swift (hunk 12 above)
[the remark this answers is not in the document]
Comment (oiuhr) [reply Tests/RulesTests+Organization.swift:643]: now on we validate it here https://github.com/nicklockwood/SwiftFormat/pull/1736#discussion_r1680318184

File: Sources/FormattingHelpers.swift (hunk 10 above)
Comment (oiuhr) [reply Sources/FormattingHelpers.swift:2014]: something stopped me from separating it to separate arrays (this will be v:v + v:t + t:v + t:t and i guess its just not so elegant), but for most of the other comments i guess we're okay for now 

maybe should wrote a few more test cases but I didn't want to overthink until full feature design review passed 🫣

**Review by calda — COMMENTED** [review]
This is REALLY AWESOME! Nice work! I really like the way you implemented the validation. Great idea!

I left a few minor comments, but otherwise this is looking really good!!

File: Rules.md [hunk 13]
```diff
@@ -1369,10 +1369,26 @@ Option | Description
...
+`--visibilityorder` | Order for visibility groups inside declaration
+`--typeorder` | Order for declaration type groups inside declaration
+`--visibilitymarks` | Marks for concrete visibility groups (public:Public Fields,..)
+`--typemarks` | Marks for concrete declaration type groups (classMethod:Baaz,..)
 
 <details>
 <summary>Examples</summary>
 
+Default value for `--visibilityorder`:
+`["beforeMarks", "instanceLifecycle", "open", "public", "package", "internal", "fileprivate", "private"]`
```
[the remark this answers is not in the document]
Comment (oiuhr) [reply Rules.md]: yeah, thanks! this looks much better

File: Tests/CommandLineTests.swift [hunk 14]
```diff
@@ -262,7 +262,7 @@ class CommandLineTests: XCTestCase {
     func testHelpLineLength() {
         CLI.print = { message, _ in
             for line in message.components(separatedBy: "\n") {
-                XCTAssertLessThanOrEqual(line.count, 80, line)
+                XCTAssertLessThanOrEqual(line.count, 100, line)
```
[the remark this answers is not in the document]
Comment (oiuhr) [reply Tests/CommandLineTests.swift]: sry, rolled back

File: Tests/RulesTests+Organization.swift [hunk 15]
```diff
@@ -163,18 +163,18 @@ class OrganizationTests: RulesTests {
...
-            func foo() -> Foo {
-                Foo()
-            }
+            var a = ""
 
             override func bar() -> Bar {
                 Bar()
             }
 
+            func foo() -> Foo {
```
[the remark this answers is not in the document]
Comment (oiuhr) [reply Tests/RulesTests+Organization.swift:174]: I feel like shuffling it up a bit by making current state the default one hurts nobody. In my personal opinion if we'll provide a different default order arguments for different modes it may become a bit messy (like features (sorting groups) may just vanish into implementation, or suddenly appear after switching to another mode) while using defaults with default configuration. although it's relatively easy to fix by mentioning all of this, including every `category` (as _what is `overridenFunctions`_), in `Example.swift` 🤧 

but my opinion is not a judgement for this case anyway, feel free to make a decision as a maintainer 👍 i guess newcomers will override config files anyway so it's a matter of documentation

File: Tests/RulesTests+Organization.swift [hunk 16]
```diff
@@ -253,14 +253,14 @@ class OrganizationTests: RulesTests {
...
 
-            func foo() -> Foo {
-                Foo()
-            }
-
             func bar() -> some View {
                 EmptyView()
             }
 
+            func foo() -> Foo {
```
[the remark this answers is not in the document]
Comment (oiuhr) [reply Tests/RulesTests+Organization.swift:260]: guess its because of https://github.com/nicklockwood/SwiftFormat/pull/1736#discussion_r1681390856

File: Tests/RulesTests+Organization.swift (hunk 15 above)
[the remark this answers is not in the document]
Comment (calda) [reply Tests/RulesTests+Organization.swift:174]: Makes sense to me 👍🏻

**Review by calda — APPROVED** [review]
Really great work! This is awesome! Users are going to love it.
